### PR TITLE
Add FileSystemToolset for file system access

### PR DESCRIPTION
**Pull request by DEENUU1**
## Summary

Add new `FileSystemToolset` that allows agents to access the local file system with security controls.

### Features
- **Operations**: `read`, `write`, `list`, `info`, `search`, `delete`
- **Security controls**:
  - `allowed_paths` - restrict access to specific directories
  - `allow_write` / `allow_delete` - control write/delete permissions
  - `max_file_size` - limit readable file size
  - `allowed_extensions` - filter by file extensions
- Path traversal protection (prevents `../` attacks)

### Usage

```python
from pydantic_ai import Agent
from pydantic_ai.toolsets import FileSystemToolset

agent = Agent(
    'openai:gpt-4.1',
    toolsets=[
        FileSystemToolset(
            allowed_paths=['/home/user/data'],
            allow_write=True,
            allowed_extensions=['.txt', '.json'],
        ),
    ],
)
```

Also available as `FileSystemTool` builtin tool configuration:

```python
from pydantic_ai import Agent
from pydantic_ai.builtin_tools import FileSystemTool

agent = Agent(
    'openai:gpt-4.1',
    builtin_tools=[
        FileSystemTool(
            allowed_paths=['/home/user/data'],
        ),
    ],
)
```

## Test plan

- [x] Path validation tests (allowed paths, traversal attacks, cwd restriction)
- [x] Extension filtering tests
- [x] Operation permission tests (read, write, delete)
- [x] File size limit tests
- [x] All file operations (read, write, list, info, search, delete)
- [x] Tool definition generation tests
- [x] Edge cases and error handling
- [x] ToolManager integration test
- [x] 100% coverage for `filesystem.py`

---

Context: Found during production work at [Vstorm](https://vstorm.co) · [github.com/vstorm-co](https://github.com/vstorm-co).

### CODE REVIEW

File: pydantic_ai_slim/pydantic_ai/builtin_tools.py [hunk 1]
```diff
@@ -400,3 +405,70 @@ def _tool_discriminator(tool_data: dict[str, Any] | AbstractBuiltinTool) -> str:
         return tool_data.get('kind', AbstractBuiltinTool.kind)
     else:
         return tool_data.kind
+
+
+@dataclass(kw_only=True)
+class FileSystemTool(AbstractBuiltinTool):
+    """A builtin tool that allows your agent to access the file system.
+
+    Unlike other builtin tools that are implemented by model providers,
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/builtin_tools.py:414]: Built-in tools are really server tools, so I don't think we should have this here and users can just create the `FileSystemToolset` directly.

File: pydantic_ai_slim/pydantic_ai/toolsets/filesystem.py [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+            path: str
+
+        class FileSystemArgs(FileSystemArgsRequired, total=False):
+            content: str
+            pattern: str
+
+        validator = cast(SchemaValidatorProt, TypeAdapter(FileSystemArgs).validator)
+
+        return {
+            'file_system': ToolsetTool(
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/toolsets/filesystem.py:405]: Why one tool with operations, instead of separate tools?

**Review by DouweM — CHANGES_REQUESTED** [review]
@DEENUU1 For handy toolsets like this that are not exactly core to Pydantic AI, what do you think about including them vs having them as separate packages we can link to from the docs? The latter would keep our maintenance burden a little more manageable, and I think file system stuff can be opinionated enough (especially in terms of what requires approval) that different users may have different needs, and we could end up with tons of issues and contributions that'd distract us from the core framework, and would be better addressed on a separate project that can grow as complicated as it wants to.

File: pydantic_ai_slim/pydantic_ai/builtin_tools.py (hunk 1 above)
Comment (DEENUU1) [reply pydantic_ai_slim/pydantic_ai/builtin_tools.py:414]: Thanks for the clarification! 

That makes total sense. I have a quick question: for community-built toolsets like this, do you prefer to host them under the Pydantic AI organization, or is the idea that contributors maintain them in their own repos and then you'd link to them from the docs? 
Also, is there (or will there be) some kind of process for submitting community packages to be listed in the documentation?

On a side note, I noticed on Slack that there's already a package for this: https://pypi.org/project/pydantic-ai-filesystem-sandbox/. So there's probably no point in duplicating the effort here. I'll close this PR.

Comment (DouweM) [reply pydantic_ai_slim/pydantic_ai/builtin_tools.py:414]: > That makes total sense. I have a quick question: for community-built toolsets like this, do you prefer to host them under the Pydantic AI organization, or is the idea that contributors maintain them in their own repos and then you'd link to them from the docs?

@DEENUU1 The latter; we can only have so much maintenance bandwidth so we'd love to have the community help build out the ecosystem of useful toolset setc.

> Also, is there (or will there be) some kind of process for submitting community packages to be listed in the documentation?

No particular process, a PR to this section will do: https://ai.pydantic.dev/toolsets/#third-party-toolsets

> On a side note, I noticed on Slack that there's already a package for this: [pypi.org/project/pydantic-ai-filesystem-sandbox](https://pypi.org/project/pydantic-ai-filesystem-sandbox/). So there's probably no point in duplicating the effort here. I'll close this PR.

Good call; did you check if that one will work for your use case?

I've asked the author on Slack to add it to our docs.
